### PR TITLE
HLasse/docs-add-docstring-to-all-user-facing-specs

### LINKED ISSUE #58: docs: add docstring to all user facing specs
Document all attributes of user facing Specs (and their parents).

Do as the last thing, as will suck if we change any names

See https://github.com/mansenfranzen/autodoc_pydantic and https://blog.nicobako.dev/articles/pydantic_autodoc.html for potentially clever ways of doing this

Might be fruitful to combine the parser from `autodoc_pydantic` with the method from the blog -- this class here https://github.com/sphinx-doc/sphinx/blob/cc314f13e8a98393ab018d83d8957a724a6f338a/sphinx/pycode/parser.py#L531

### PR DESCRIPTION
**Pull request by HLasse**
- [ ] I have assigned ranges (e.g. `>=0.1, <0.2`) to all new dependencies (allows dependabot to keep dependency ranges wide for better compatability)

Fixes #58

## Notes for reviewers
Reviewers can skip X, but should pay attention to Y.

### CODE REVIEW

File: src/timeseriesflattener/feature_spec_objects.py [hunk 1]
```diff
@@ -98,6 +105,27 @@ class Config:
...
+        field_info: FieldInfo = field.field_info
+        doc += f"{name} ({type_}):\n        "
+        default_value = field.default
+        default_str = (
+            f"Defaults to: {default_value}." if default_value is not None else ""
+        )
+        doc += f"    {field_info.description} {default_str}\n"
+    # remove the last newline
+    doc = doc[:-1]
+    return doc
```
Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:126]: This was very hard to parse for me. Some questions:

0. The list comprehension is very hard to read, unless you already know the exact underlying data strcture. Can we refactor it in some way, so there is less indexing? (Extra variables?)
1. Name is the attribute name? Recommend `field_name`.
2. Field is the field object? Maybe call it `field_obj`?

I think I'd prefer extracting relevant information first, then adding to the docstring. 

Makes it easier to first parse the logic, then the formatting.

File: src/timeseriesflattener/feature_spec_objects.py [hunk 2]
```diff
@@ -124,33 +152,76 @@ def check_that_col_names_in_kwargs_exist_in_df(data: dict[str, Any], df: pd.Data
...
-
-    output_col_name_override: Optional[str] = None
-    # Override the generated col name after flattening the time series.
+    Fields:
+        values_loader (Optional[Callable]):
+            Loader for the df. Tries to resolve from the data_loaders registry,
+            then calls the function which should return a dataframe.
+        values_name (Optional[str]):
+            A string that corresponds to a key in a dictionary of multiple
+            dataframes that corresponds to a name of a type of values.
```
Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:161]:
```suggestion
            A string that maps to a key in a dictionary instantiated by `split_df_and_register_to_dict`.
            Each key corresponds to a dataframe, which is a subset of the df where the values_name == key.
```

File: src/timeseriesflattener/feature_spec_objects.py [hunk 3]
```diff
@@ -203,38 +274,101 @@ def __eq__(self, other):
...
-
-    entity_id_col_name: str = "entity_id"
-    # Col name for ids in the input dataframe.
-
-    loader_kwargs: Optional[dict] = None
-
-    # Optional keyword arguments for the data loader
+    """The minimum specification required for all
+        collapsed time series. (temporal features), whether looking ahead or behind.
+        Mostly used for inheritance below.
```
Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:284]:
```suggestion
    """The minimum specification required for collapsing a temporal, whether looking ahead or behind.
        Mostly used for inheritance below.
```

File: src/timeseriesflattener/feature_spec_objects.py [hunk 4]
```diff
@@ -203,38 +274,101 @@ def __eq__(self, other):
...
+
+    Fields:
+        values_loader (Optional[Callable]):
+            Loader for the df. Tries to resolve from the data_loaders registry,
+            then calls the function which should return a dataframe.
+        values_name (Optional[str]):
+            A string that corresponds to a key in a dictionary of multiple
+            dataframes that corresponds to a name of a type of values.
```
Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:292]:
```suggestion
        values_name (Optional[str]):
            A string that maps to a key in a dictionary instantiated by `split_df_and_register_to_dict`.
            Each key corresponds to a dataframe, which is a subset of the df where the values_name == key.
```

File: src/timeseriesflattener/feature_spec_objects.py [hunk 5]
```diff
@@ -203,38 +274,101 @@ def __eq__(self, other):
...
+        allowed_nan_value_prop (float):
+            If NaN is higher than this in the input dataframe during
+            resolution, raise an error. Defaults to: 0.0.
+        entity_id_col_name (str):
+            Col name for ids in the input dataframe. Defaults to: entity_id."""
+
+    class Doc:
+        short_description = """The minimum specification required for all 
+        collapsed time series. (temporal features), whether looking ahead or behind.
+        Mostly used for inheritance below."""
```
Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:332]:
```suggestion
        short_description = """The minimum specification required for temporal time series."""
```

Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:332]:
```suggestion
    class Doc:
        short_description = """The minimum specification required for all 
        collapsed time series. (temporal features), whether looking ahead or behind.
        Mostly used for inheritance below."""
```

When is the `Doc` class used?

File: src/timeseriesflattener/feature_spec_objects.py (hunk 2 above)
Comment (MartinBernstorff) [reply src/timeseriesflattener/feature_spec_objects.py:161]: Cmd+d to modify this in the entire document. So much docs duplication D:

File: src/timeseriesflattener/feature_spec_objects.py [hunk 6]
```diff
@@ -89,7 +91,12 @@ def resolve_values_df(data: dict[str, Any]):
 
 
 class BaseModel(PydanticBaseModel):
-    """."""
+    """Modified Pydantic BaseModel to allow arbitrary
+    types and disallow attributes not in the class."""
```
Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:95]: Allegedly, Sphinx already pulls parent docstrings. When merging, can you check if that's the case in the docs? If so, we'll have to disable it in Sphinx.

https://stackoverflow.com/questions/2025562/inherit-docstrings-in-python-class-inheritance

File: src/timeseriesflattener/feature_spec_objects.py [hunk 7]
```diff
@@ -285,13 +475,79 @@ def __init__(self, **data):
...
+            <prefix>_<outcome_name>.""",
+    )
 
-    incident: bool
+    incident: bool = Field(
+        description="""Whether the outcome is incident or not, i.e. whether you 
+            can experience it more than once. For example, type 2 diabetes is incident. 
+            Incident outcomes can be handled in a vectorised way during resolution, 
+            which is faster than non-incident outcomes.""",
+    )
```
Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:546]:
```suggestion
    incident: bool = Field(
        description="""Whether the outcome is incident or not. 
            I.e., incident outcomes are outcomes you can only experience once. 
                For example, type 2 diabetes is incident. 
            Incident outcomes can be handled in a vectorised way during resolution, 
            which is faster than non-incident outcomes.""",
    )
```

File: src/timeseriesflattener/feature_spec_objects.py [hunk 8]
```diff
@@ -326,38 +578,59 @@ def is_dichotomous(self) -> bool:
 
 
 class MinGroupSpec(BaseModel):
```
Comment (MartinBernstorff) [inline src/timeseriesflattener/feature_spec_objects.py:580]:
```suggestion
class _MinGroupSpec(BaseModel):
```

Perhaps we should make this, `MinSpec` and `AnySpec` private?

File: tests/test_timeseriesflattener/test_feature_spec_objects.py [hunk 9]
```diff
@@ -133,3 +141,57 @@ def test_resolve_multiple_fn_to_str():
...
+
+    if current_docstring_no_whitespace != generated_docstring_no_whitespace:
+        raise AssertionError(
+            f"""{spec} docstring is not updated correctly.
+        Got: \n\n{current_docstring}.
+        Expected: \n\n{generated_docstring}
+        \n\nDiff:
+        {pretty_diff(current_docstring, generated_docstring)}
+        """,
+        )
```
Comment (MartinBernstorff) [inline tests/test_timeseriesflattener/test_feature_spec_objects.py:197]: Document intended behaviour from the user on test failure.

**Review by MartinBernstorff — CHANGES_REQUESTED** [review]

File: src/timeseriesflattener/feature_spec_objects.py (hunk 5 above)
Comment (HLasse) [reply src/timeseriesflattener/feature_spec_objects.py:332]: We also talked about this yesterday, but for future reference. The `Doc` class is used to generate the top part (the description) of the docstring. Will add a note on how to modify the docstrings correctly

File: src/timeseriesflattener/feature_spec_objects.py (hunk 6 above)
Comment (HLasse) [reply src/timeseriesflattener/feature_spec_objects.py:95]: Will do!

**Review by MartinBernstorff — APPROVED** [review]
Looks great! Tiny addition from me. Thanks a ton for this!
